src/main/java/org/elasticsearch/index/query/QueryBuilder.java

```diff
@@ -45,4 +45,10 @@
      * @throws IOException
      */
     Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException;
+
+    /**
+     * Validate the query.
+     * @throws ElasticsearchException if the query is not valid, e.g. fields that are needed to create the lucene query are missing.
+     */
+    void validate() throws ElasticsearchException;
```

Comment:
Looked at ActionRequestValidationException, could make a copy of that as QueryValidationException. It seems like they are never thrown though, rather returned, so signature would be `QueryValidationException validate()` then? Could also collect all errors and then throw exception at the end.